src/install/lockfile.zig

```diff
@@ -2522,25 +2629,71 @@ pub const Package = extern struct {
                 out_group_i += 1;
             }
 
+            if (features.workspaces) {
+                out_groups[out_group_i] = DependencyGroup.workspaces;
+                out_group_i += 1;
+            }
+
             break :brk out_groups;
         };
 
+        var workspace_names: []string = undefined;
         inline for (dependency_groups) |group| {
             if (json.asProperty(group.prop)) |dependencies_q| {
-                if (dependencies_q.expr.data == .e_object) {
-                    for (dependencies_q.expr.data.e_object.properties.slice()) |item| {
-                        const key = item.key.?.asString(allocator) orelse "";
-                        const value = item.value.?.asString(allocator) orelse "";
+                switch (dependencies_q.expr.data) {
+                    .e_array => |arr| {
+                        workspace_names = try allocator.alloc(string, arr.items.len);
+
+                        var workspace_log = logger.Log.init(allocator);
+                        defer log.deinit();
+
+                        var workspace_buf = try allocator.alloc(u8, 1024);
+                        defer allocator.free(workspace_buf);
+
+                        for (arr.slice()) |item, i| {
+                            const path = item.asString(allocator) orelse return error.InvalidPackageJSON;
+
+                            var workspace_dir = try std.fs.cwd().openDir(path, .{});
+                            defer workspace_dir.close();
 
-                        string_builder.count(key);
-                        string_builder.count(value);
+                            var workspace_file = try workspace_dir.openFile("package.json", .{ .mode = .read_only });
+                            defer workspace_file.close();
 
-                        // If it's a folder, pessimistically assume we will need a maximum path
-                        if (Dependency.Version.Tag.infer(value) == .folder) {
+                            const workspace_read = try workspace_file.preadAll(workspace_buf, 0);
+                            const workspace_source = logger.Source.initPathString(path, workspace_buf[0..workspace_read]);
+
+                            initializeStore();
+
+                            var workspace_json = try json_parser.PackageJSONVersionChecker.init(allocator, &workspace_source, &workspace_log);
+
+                            _ = try workspace_json.parseExpr();
+                            if (!workspace_json.has_found_name) return error.InvalidPackageJSON;
+
+                            const workspace_name = workspace_json.found_name;
+
+                            string_builder.count(workspace_name);
+                            string_builder.count(path);
                             string_builder.cap += bun.MAX_PATH_BYTES;
+                            workspace_names[i] = try allocator.dupe(u8, workspace_name);
```

Comment:
this should be using the string_builder stuff instead of a separate memory allocation

Reply:
I struggled with this quite a bit &minus; this section of code is still counting the `.cap` for `string_builder`, so it has not been `.allocate()`'d yet.

Any ideas how to get around that?